R/main.R

```diff
@@ -393,6 +398,57 @@ main_do_batch <- function(x) {
   message("ids:", paste(ids, collapse = ", "))
 }
 
+## 9. workflow
+usage_workflow <- "Usage:
+  orderly workflow [options] <name> [<parameter>...]
+
+Options:
+  --print-log      Print the log (rather than storing it)
+  --pull           Pull git before running report
+  --instance=NAME  Database instance to use (if instances are configured)
+  --message=TEXT   A message explaining why the workflow was run
+
+Parameters, if given, must be passed through in key=value pairs"
+
+main_do_workflow <- function(x) {
+  ## TODO: Get some classed errors though here and then write out
+  ## information about whether or not things worked and why they
+  ## didn't.  Possible issues (in order)
+  ##
+  ## * orderly report not found
+  ## * error while preparing (e.g., package not found)
+  ## * error while running report
+  ## * error checking artefacts
+  config <- orderly_config_get(x$options$root, TRUE)
+  name <- x$options$name
+  instance <- x$options$instance
+  print_log <- x$options$print_log
+  pull <- x$options$pull
+  message <- x$options$message
+
+  if (print_log) {
+    sink(stderr(), type = "output")
+    on.exit(sink(NULL, type = "output"))
+  } else {
+    config$add_run_option("capture_log", TRUE)
+  }
+
+  if (pull) {
+    git_pull(config$root)
+  }
+
+  run_reports <- orderly_workflow(name, root = config, instance = instance,
+                                  message = message)
+  output <- vcapply(run_reports, function(report) {
+    path_rds <- path_orderly_run_rds(
+      file.path(config$root, "archive", report$name, report$id))
+    post_success(readRDS(path_rds), config)
```

Comment:
Because we post the success here which needs the report name and ID I modified `orderly_workflow` to return both the id and the name of the reports it has run. That feels like a fine change to me but is now inconsistent with batch & run but as they only ever run for 1 report seems fine.

Alternative is to post something different for a workflow, i.e. post once the workflow has completed that is has run and includes reports x, y, z with ids x, y, z. I can always do that in another PR.

Reply:
I like the idea of it posting as it goes because then you do get more progress information - can we move this into the `workflow$run` method perhaps?  This feels similar to the internal- and external- facing view on orderly run that we need to think about properly at some point.

Posting something afterwards for the workflow might be nice, and we could also tweak the `post_success` to indicate that we're part of a workflow. Another thing we might want to add to orderly web is a workflow page - so we could link there. Lots of things to add! But of this set of things need doing here.

Reply:
Having a look this already is being done in `workflow_version` on a `commit` so I have added a test which checks that `post_success` gets called for each report in a workflow. Can add a ticket for expanding `post_success` to have some indication that we are part of a workflow